Source/missiles.cpp

```diff
@@ -410,19 +416,24 @@ void CheckMissileCol(Missile &missile, int minDamage, int maxDamage, bool isDama
 
 	bool isPlayerHit = false;
 	bool blocked = false;
-	const int8_t pid = dPlayer[mx][my];
+	int8_t pid = dPlayer[mx][my];
 	if (pid > 0) {
-		if (missile._micaster != TARGET_BOTH && !missile.IsTrap()) {
-			if (missile._micaster == TARGET_MONSTERS) {
-				if ((pid - 1) != missile._misource)
-					isPlayerHit = Plr2PlrMHit(Players[missile._misource], pid - 1, minDamage, maxDamage, missile._midist, missile._mitype, isDamageShifted, &blocked);
-			} else {
-				Monster &monster = Monsters[missile._misource];
-				isPlayerHit = PlayerMHit(pid - 1, &monster, missile._midist, minDamage, maxDamage, missile._mitype, isDamageShifted, 0, &blocked);
+		pid -= 1;
+		int earflag = 0;
+		switch (missileSource) {
+		case MissileSource::Player: {
+			Player *player = missile.sourcePlayer();
+			if (&Players[pid] != player && missile._micaster != TARGET_BOTH) {

```

Comment:
I'm not sure exactly how to classify this, but it looks like firewalls damages players as if they where cast by traps even when cast by players because of the special case for `TARGET_BOTH`. This PR makes this behavior more clear. I'm not sure if it should be addressed or not.

Reply:
Perhaps we should keep the behaviour as is for now and create a separate issue to handle this later?